test/matrix_multiply_add.jl

```diff
@@ -0,0 +1,178 @@
+using StaticArrays
+using LinearAlgebra
+using BenchmarkTools
+using Test
+
+macro test_noalloc(ex)
+    esc(quote
+        $ex
+        @test(@allocated($ex) == 0)
+    end)
+end
```

Comment:
If you've found `@benchmark` reliable for this, let's just make `test_noalloc` use that instead? Something like:

```
macro test_noalloc(ex)
    quote
        bmark = @benchmark $(esc(ex)) samples=10 evals=10
        @test minimum(bmark).allocs == 0
    end
end
```

You could put this in `test_util` as well and we can use it everywhere instead of `@allocated`? Couple of questions:
* We don't actually want to do a benchmark here so it seems a bit heavy weight. We can put up with that though, unless it slows down the tests a lot?
* Can you reduce the samples and evals?
* Is the `minimum` there required, or will anything other than the maximum do? (Are we measuring the right thing?)

Reply:
Ugh, there's something really broken about the way Julia or BenchmarkTools processes the correct-seeming use of `esc` in the nested macro above.

Also I didn't realize quite what `minimum` does; how about this instead:

```julia
macro test_noalloc(ex)
    ex = :(@test minimum(@benchmark($ex, samples=10, evals=10)).allocs == 0)
    ex.args[2] = __source__  # trick: make `@test` aware of the source location of `@test_noalloc`
    esc(ex)
end
```

Reply:
`@test_noalloc` macro sounds like a good idea. We can swap `@benchmark` with something else if these tests slow down the whole test suite a lot.  My guess is that it'd be OK as long as you have something like `samples=10 evals=10`.

> Is the `minimum` there required

`bmark` is the `Trial` whose `.allocs` field is already an `Int`:

https://github.com/JuliaCI/BenchmarkTools.jl/blob/cf7cada0b3f41a00c70ef03891ffae29c31f8573/src/trials.jl#L5-L11

It seems that `Base.push!(t::Trial, time, gctime, memory, allocs)` (which is used to accumulate the benchmark trial results) already takes `maximum` over all the trials:

https://github.com/JuliaCI/BenchmarkTools.jl/blob/cf7cada0b3f41a00c70ef03891ffae29c31f8573/src/trials.jl#L29

(Also, `minimum` is the minimum over the wall times. So I don't think we need it anyway.)

Reply:
Your suggested `test_noalloc(ex)` macro didn't work for me? When I use it within a function (not in global scope) it can't find the local arguments.

Reply:
Try put a `$` in front of the local arguments (it's a feature of `@benchmark` to make sure that it captures them properly).

Reply:
> Your suggested test_noalloc(ex) macro didn't work for me?

Andy's suggestion seems to work. But oh wow, I had a look at `BenchmarkTools.generate_benchmark_definition`, it's kind of a horror show :grimacing: it calls `Core.eval` in the macro and really seems to go out of its way to subvert the usual macro expansion rules in a weird way.

Maybe it has to be that way for some reason, but it sure is doing some weird stuff.

Reply:
If it's too hard to wrap `@benchmark` reliably I'm completely ok with just having it written out by hand in a few places.

I tried looking more into why `@allocated` is failing but it quickly led to a very deep rabbit hole which I'm still trying to climb out of.